src/sema/unused_variable.rs

```diff
@@ -0,0 +1,409 @@
+/// Namespace (for storage variables)
+pub fn assigned_variable(ns: &mut Namespace, exp: &Expression, symtable: &mut Symtable) {
+    match &exp {
+        Expression::StorageVariable(_, _, contract_no, offset) => {
+            ns.contracts[*contract_no].variables[*offset].assigned = true;
+        }
+
+        Expression::Variable(_, _, offset) => {
+            let var = symtable.vars.get_mut(offset).unwrap();
+            (*var).assigned = true;
+        }
+
+        Expression::StructMember(_, _, str, _) => {
+            assigned_variable(ns, str, symtable);
+        }
+
+        Expression::Subscript(_, _, array, index)
+        | Expression::DynamicArraySubscript(_, _, array, index)
+        | Expression::StorageBytesSubscript(_, array, index) => {
+            assigned_variable(ns, array, symtable);
+            used_variable(ns, index, symtable);
+        }
+
+        Expression::Trunc(_, _, var)
+        | Expression::Cast(_, _, var)
+        | Expression::BytesCast(_, _, _, var) => {
+            assigned_variable(ns, var, symtable);
+        }
+
+        _ => {}
+    }
+}
+
+/// Mark variables as used, either in the symbol table (for state variables) or in the
+/// Namespace (for global constants and storage variables)
+/// The functions handles complex expressions in a recursive fashion, such as array length call,
+/// assign expressions and array subscripts.
+pub fn used_variable(ns: &mut Namespace, exp: &Expression, symtable: &mut Symtable) {
+    match &exp {
+        Expression::StorageVariable(_, _, contract_no, offset) => {
+            ns.contracts[*contract_no].variables[*offset].read = true;
+        }
+
+        Expression::Variable(_, _, offset) => {
+            let var = symtable.vars.get_mut(offset).unwrap();
+            (*var).read = true;
+        }
+
+        Expression::ConstantVariable(_, _, Some(contract_no), offset) => {
+            ns.contracts[*contract_no].variables[*offset].read = true;
+        }
+
+        Expression::ConstantVariable(_, _, None, offset) => {
+            ns.constants[*offset].read = true;
+        }
+
+        Expression::StructMember(_, _, str, _) => {
+            used_variable(ns, str, symtable);
+        }
+
+        Expression::Subscript(_, _, array, index)
+        | Expression::DynamicArraySubscript(_, _, array, index)
+        | Expression::StorageBytesSubscript(_, array, index) => {
+            used_variable(ns, array, symtable);
+            used_variable(ns, index, symtable);
+        }
+
+        Expression::Assign(_, _, assigned, assignee) => {
+            assigned_variable(ns, assigned, symtable);
+            used_variable(ns, assignee, symtable);
+        }
+
+        Expression::DynamicArrayLength(_, array)
+        | Expression::StorageArrayLength {
+            loc: _,
+            ty: _,
+            array,
+            ..
+        } => {
+            used_variable(ns, array, symtable);
+        }
+
+        Expression::StorageLoad(_, _, expr)
```

Comment:
This function only recurses for limited set of expressions. For example, a simple `Expression::Add` is not mentioned. If you want to proper recursion there is a `recurse()` method on `Expression`.

However, as I read the code, I don't think this function needs to be recurse at all. sema resursively resolves expressions already.

Reply:
Sema does resolve expressions recursively, as you mentioned. The problem is that using only `Expression::Variable` (and its variants) does not provide me the context in which the variable has been used, i.e., I cannot differentiate between an assignment and a read. Once I target that, I'll need to recurse other expressions that encapsulate a variable, such as `StorageLoad` and `ZeroExt` (but not limited to these two cases).

For `Expression::Add` and other expressions that have a left and a right side, I use the function `check_var_usage_expression`.

I agree that this function is worth some refactoring. I am happy to discuss more during our meeting tomorrow.

Reply:
I simplified the function. Please, take a look again. I tried to recurse only for expressions that directly wrap a variable.